contracts/token/ERC721/extensions/ERC721Consecutive.sol

```diff
@@ -0,0 +1,116 @@
+        if (owner != address(0) || tokenId > type(uint96).max) {
+            return owner;
+        }
+
+        // Otherwise, check the token was not burned, and fetch ownership from the anchors
+        // Note: no need for safe cast, we know that tokenId <= type(uint96).max
+        return _sequentialBurn.get(tokenId) ? address(0) : address(_sequentialOwnership.lowerLookup(uint96(tokenId)));
+    }
+
+    /**
+     * @dev Mint a batch of token of length `batchSize` for `to`.
+     *
+     * WARNING: Consecutive mint is only available during construction. ERC721 requires that any minting done after
+     * construction emits a Transfer event, which is not the case of mints performed using this function.
+     *
+     * WARNING: Consecutive mint is limited to batches of 5000 tokens. Further minting is possible from a contract's
+     * point of view but would cause indexing issues for off-chain services.
+     *
+     * Emits a {ConsecutiveTransfer} event.
+     */
+    function _mintConsecutive(address to, uint96 batchSize) internal virtual {
+        // minting a batch of size 0 is a no-op
+        if (batchSize == 0) return;
+
+        require(!Address.isContract(address(this)), "ERC721Consecutive: batch minting restricted to constructor");
+        require(to != address(0), "ERC721Consecutive: mint to the zero address");
+        require(batchSize < 5000, "ERC721Consecutive: batches too large for indexing");
+
+        uint96 first = _totalConsecutiveSupply();
+        uint96 last = first + batchSize - 1;
+
+        // hook before
+        _beforeConsecutiveTokenTransfer(address(0), to, first, last);
+
+        // push an ownership checkpoint & emit event
+        _sequentialOwnership.push(SafeCast.toUint96(last), uint160(to));
+        emit ConsecutiveTransfer(first, last, address(0), to);
+
+        // hook after
+        _afterConsecutiveTokenTransfer(address(0), to, first, last);
+    }
+
+    /**
+     * @dev See {ERC721-_mint}. Override version that restricts normal minting to after construction.
+     *
+     * Warning: Using {ERC721Consecutive} prevents using {_mint} during construction in favor of {_mintConsecutive}.
+     * After construction, {_mintConsecutive} is no longer available and {_mint} becomes available.
+     */
+    function _mint(address to, uint256 tokenId) internal virtual override {
+        require(Address.isContract(address(this)), "ERC721Consecutive: can't mint during construction");
+
+        super._mint(to, tokenId);
+        if (_sequentialBurn.get(tokenId)) {
+            _sequentialBurn.unset(tokenId);
+        }
+    }
+
+    /**
+     * @dev See {ERC721-_mint}. Override needed to avoid looking up in the sequential ownership structure for burnt
+     * token.
+     */
+    function _burn(uint256 tokenId) internal virtual override {
+        super._burn(tokenId);
+        if (tokenId <= _totalConsecutiveSupply()) {
+            _sequentialBurn.set(tokenId);
```

Comment:
As usual we have a problem with the hooks here. `_afterTokenTransfer` will be invoked in a state where the owner is still set. An alternative would be to add this code as an override to `_afterTokenTransfer` but that doesn't fully fix things because some other overrides that execute prior may still get invoked in that state.